test/python/primitives/test_primitive.py

```diff
@@ -110,3 +114,48 @@ def test_value_error(self):
         """Test value error if no parameter_values or default are provided."""
         with self.assertRaises(ValueError):
             BasePrimitive._validate_parameter_values(None)
+
+
+class TestCircuitKey(QiskitTestCase):
+    """Tests for _circuit_key function"""
+
+    def test_different_circuits(self):
+        """Test collision of quantum circuits."""
+
+        with self.subTest("Ry circuit"):
+
+            def test_func(n):
+                qc = QuantumCircuit(1, 1, name="foo")
+                qc.ry(n, 0)
+                return qc
+
+            keys = [_circuit_key(test_func(i)) for i in range(5)]
+            self.assertEqual(len(keys), len(set(keys)))
+
+        with self.subTest("pulse circuit"):
+
+            def test_with_scheduling(n):
+                custom_gate = pulse.Schedule(name="custom_x_gate")
+                custom_gate.insert(
+                    0, pulse.Play(pulse.Constant(160 * n, 0.1), pulse.DriveChannel(0)), inplace=True
+                )
+                qc = QuantumCircuit(1)
+                qc.x(0)
+                qc.add_calibration("x", qubits=(0,), schedule=custom_gate)
+                return transpile(qc, FakeAlmaden(), scheduling_method="alap")
+
+            keys = [_circuit_key(test_with_scheduling(i)) for i in range(1, 5)]
+            self.assertEqual(len(keys), len(set(keys)))
+
+    def test_circuit_key_controlflow(self):
+        """Test for a circuit with control flow."""
+        qc = QuantumCircuit(2, 1)
+
+        with qc.for_loop(range(5)):
+            qc.h(0)
+            qc.cx(0, 1)
+            qc.measure(0, 0)
+            qc.break_loop().c_if(0, True)
+
+        self.assertIsInstance(hash(_circuit_key(qc)), int)
+        self.assertIsInstance(json.dumps(_circuit_key(qc)), str)
```

Comment:
This `json` line doesn't look like it belongs in Terra - JSON serialisation of the internal object `_circuit_key` shouldn't be something Terra needs, and no package downstream of us should be relying on a private function having any particular behaviour.

Using a dummy primitive whose `_run` method (or whatever) just makes assertions / leaks the received circuit back to the caller somehow could be a cleaner way of making a public-API test.